src/linear_algebra/matrix/diag.lean

```diff
@@ -0,0 +1,210 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import data.matrix.kronecker
+
+/-!
+# Diagonal matrices
+
+This file contains the definition and basic results about diagonal matrices.
+
+## Main results
+
+- `matrix.has_orthogonal_rows`:
+  `A.has_orthogonal_rows` means `A` has orthogonal (with respect to `dot_product`) rows.
+- `matrix.has_orthogonal_cols`:
+  `A.has_orthogonal_cols` means `A` has orthogonal (with respect to `dot_product`) columns.
+- `matrix.is_diag`: a proposition that stats a given square matrix `A` is diagonal
+                    (i.e. `∀ m n, i ≠ j → A m n = 0`).
+
+## Tags
+
+diag, diagonal, matrix
+-/
+
+namespace matrix
+
+variables {α β R n m : Type*}
+
+open function
+open_locale matrix kronecker
+
+/-- `A.has_orthogonal_rows` means matrix `A` has orthogonal
+    (with respect to `dot_product`) rows. -/
+def has_orthogonal_rows [fintype n] [has_mul α] [add_comm_monoid α]
+(A : matrix m n α) : Prop :=
+∀ ⦃i₁ i₂⦄, i₁ ≠ i₂ → dot_product (A i₁) (A i₂) = 0
+
+/-- `A.has_orthogonal_cols` means matrix `A` has orthogonal
+    (with respect to `dot_product`) columns. -/
+def has_orthogonal_cols [fintype m] [has_mul α] [add_comm_monoid α]
+(A : matrix m n α) : Prop :=
+∀ ⦃i₁ i₂⦄, i₁ ≠ i₂ → dot_product (λ j, A  j i₁) (λ j, A j i₂) = 0
```

Comment:
This spelling feels quite ugly to me, and I don't think we use it elsewhere already. I think we should either prefer:
```suggestion
∀ ⦃i₁ i₂⦄, i₁ ≠ i₂ → dot_product (Aᵀ i₁) (Aᵀ i₂) = 0
```
Which we can compress to
```
has_orthogonal_rows Aᵀ
```